File: ruby/house/house.rb
# Adapted from
# https://github.com/alxndr/exercism/blob/master/ruby/house/house.rb

class House

  VERSE_MAPPING = {
    12=> {subject: 'horse and the hound and the horn', verb: 'belonged to' },
    11=> {subject: 'farmer sowing his corn',           verb: 'kept' },
    10=> {subject: 'rooster that crowed in the morn',  verb: 'woke' },
    9 => {subject: 'priest all shaven and shorn',      verb: 'married' },
    8 => {subject: 'man all tattered and torn',        verb: 'kissed' },
    7 => {subject: 'maiden all forlorn',               verb: 'milked' },
    6 => {subject: 'cow with the crumpled horn',       verb: 'tossed' },
    5 => {subject: 'dog',                              verb: 'worried' },
    4 => {subject: 'cat',                              verb: 'killed' },
    3 => {subject: 'rat',                              verb: 'ate' },
    2 => {subject: 'malt',                             verb: 'lay in the house'\
                                                             ' that Jack built'}
  }

  def self.verse(num)
    return "This is the house that Jack built.\n" if num == 1

    lines = num.downto(2).map do |n|
      verse_parts = VERSE_MAPPING[n]
      "#{verse_parts[:subject]}\nthat #{verse_parts[:verb]}"
    end

    "This is the #{lines.join(' the ')}.\n"
  end

  def self.verses(from, to)
    ( from.upto(to).map { |n| self.verse(n) }.join("\n"))
  end

  def self.recite
    verses(1,12)
  end

end

File: go/house/house.go
package house

import (
	"fmt"
	"strings"
)

func Verse(v int) string {
	if v == 1 {
		return `This is the house that Jack built.`
	}
	if v == 2 {
		return `This is the malt
that lay in the house that Jack built.`
	}
	if v == 3 {
		return `This is the rat
that ate the malt
that lay in the house that Jack built.`
	}
	if v == 4 {
		return `This is the cat
that killed the rat
that ate the malt
that lay in the house that Jack built.`
	}
	if v == 5 {
		return `This is the dog
that worried the cat
that killed the rat
that ate the malt
that lay in the house that Jack built.`
	}
	if v == 6 {
		return `This is the cow with the crumpled horn
that tossed the dog
that worried the cat
that killed the rat
that ate the malt
that lay in the house that Jack built.`
	}
	if v == 7 {
		return `This is the maiden all forlorn
that milked the cow with the crumpled horn
that tossed the dog
that worried the cat
that killed the rat
that ate the malt
that lay in the house that Jack built.`
	}
	if v == 8 {
		return `This is the man all tattered and torn
that kissed the maiden all forlorn
that milked the cow with the crumpled horn
that tossed the dog
that worried the cat
that killed the rat
that ate the malt
that lay in the house that Jack built.`
	}
	if v == 9 {
		return `This is the priest all shaven and shorn
that married the man all tattered and torn
that kissed the maiden all forlorn
that milked the cow with the crumpled horn
that tossed the dog
that worried the cat
that killed the rat
that ate the malt
that lay in the house that Jack built.`
	}
	if v == 10 {
		return `This is the rooster that crowed in the morn
that woke the priest all shaven and shorn
that married the man all tattered and torn
that kissed the maiden all forlorn
that milked the cow with the crumpled horn
that tossed the dog
that worried the cat
that killed the rat
that ate the malt
that lay in the house that Jack built.`
	}
	if v == 11 {
		return `This is the farmer sowing his corn
that kept the rooster that crowed in the morn
that woke the priest all shaven and shorn
that married the man all tattered and torn
that kissed the maiden all forlorn
that milked the cow with the crumpled horn
that tossed the dog
that worried the cat
that killed the rat
that ate the malt
that lay in the house that Jack built.`
	}
	if v == 12 {
		return `This is the horse and the hound and the horn
that belonged to the farmer sowing his corn
that kept the rooster that crowed in the morn
that woke the priest all shaven and shorn
that married the man all tattered and torn
that kissed the maiden all forlorn
that milked the cow with the crumpled horn
that tossed the dog
that worried the cat
that killed the rat
that ate the malt
that lay in the house that Jack built.`
	}
	panic(fmt.Sprintf("unsupported verseNumber %v", v))
}

func Song() string {
	verses := []string{}
	for i := 1; i <= 12; i++ {
		verse := Verse(i)
		verses = append(verses, verse)
	}
	return strings.Join(verses, "\n\n")
}
